### ISSUE TITLE
about evaluation sort part question

### ISSUE THREAD

**Issue by ideafang** [TimDettmers/ConvE#71]
# question code
```python
# sort and rank
max_values, argsort1 = torch.sort(pred1, 1, descending=True)
max_values, argsort2 = torch.sort(pred2, 1, descending=True)
...
rank1 = np.where(argsort1[i]==e2[i, 0].item())[0][0]
rank2 = np.where(argsort2[i]==e1[i, 0].item())[0][0]
```
# question
Hey, thank you for your greate work! I got some question when rewriting your code.  the above code is located in evaluation.py line 59. I think the rank1 may be calculated by `rank1 = argsort1[i][e2[i, 0].item()]`. Because `argsort1` is the array about rank.(e.g. argsort1[i][j] is i-th sample 'j' entity's rank_number,) and rank1 is want to get e2's rank_number. 

But what puzzles me is that I can’t explain why your original code works.

Hope to get your answer, thanks.

**Comment by TimDettmers** [comment]
Sorry about the confusing code. When I was working on the project I needed to get results quickly which resulted in terrible code. I never got around cleaning it up so it is less confusing.

The variable `pred1` contains the scores for all `e1` and the matrix has the score for a particular entity at index `k` where `e2[i, 0] == k`. As such, when we sort, we get the ranking of original indices. So if entity 137 scores the top rank, then `argsort1[i, 0] == 137`. Since we are looking for the rank of the entity at position `k`, we need to find the rank of the index with the value `k` and you get this via `np.where(argsort[i]==k)` where `k==e2[i, 0]`. 

Basically what we are having is a ranking of indices and not a ranking of scores. To get the ranking of scores we need to index the ranking of indices. It is very confusing and I am not proud of this piece of software, but I hope this clears up what is going on.

**Comment by ideafang** [comment]
Thank you for your answer! I didn't comprehend the `torch.sort` function before. Through your answer, I fully understand this part of the code. 
By reading your code, I can really get started with knowledge graph link prediction. Good job!
